ui/desktop/src/extensions.ts

```diff
@@ -0,0 +1,207 @@
+export type FullExtensionConfig = ExtensionConfig & {
+  id: string;
+  name: string;
+  description: string;
+  enabled: boolean;
+}
+
+// Store extension config in user_settings
+const storeExtensionConfig = (config: FullExtensionConfig) => {
+  try {
+    const userSettingsStr = localStorage.getItem('user_settings');
+    const userSettings = userSettingsStr ? JSON.parse(userSettingsStr) : { models: [], extensions: [] };
+
+    // Check if config already exists (based on cmd for stdio, uri for sse, name for builtin)
+    const extensionExists = userSettings.extensions.some(
+      (extension: { id: string }) => extension.id === config.id
+    );
+
+    if (!extensionExists) {
+      userSettings.extensions.push(config);
+      localStorage.setItem('user_settings', JSON.stringify(userSettings));
+      console.log('Extension config stored successfully in user_settings');
+    } else {
+      console.log('Extension config already exists in user_settings');
+    }
+  } catch (error) {
+    console.error('Error storing extension config:', error);
+  }
+};
+
+// Load stored extension configs from user_settings
+export const loadStoredExtensionConfigs = async (): Promise<void> => {
+  try {
+    const userSettingsStr = localStorage.getItem('user_settings');
+
+    if (userSettingsStr) {
+      const userSettings = JSON.parse(userSettingsStr);
+      const enabledExtensions = userSettings.extensions.filter((ext: any) => ext.enabled);
+
+      for (const ext of enabledExtensions) {
+        // Convert extension format back to ExtensionConfig
+        const config: ExtensionConfig = {
+          type: 'stdio', // Assuming all stored extensions are stdio type for now
+          cmd: ext.command,
+          args: [],
+          env_keys: ext.environmentVariables?.map((env: any) => env.name) || []
+        };
+
+        await extendGoosed(config);
+      }
+
+      console.log('Loaded stored extension configs from user_settings and activated extensions with agent');
+    }
+  } catch (error) {
+    console.error('Error loading stored extension configs:', error);
+  }
+};
+
+// Update the path to the binary based on the command
+export const replaceWithShims = async (cmd: string): Promise<string> => {
+  
+    const binaryPathMap: Record<string, string> = {
+      'goosed': await window.electron.getBinaryPath('goosed'),
+      'npx': await window.electron.getBinaryPath('npx'),
+      'uvx': await window.electron.getBinaryPath('uvx'),
+    };    
+    
+    if (binaryPathMap[cmd]) {
+      console.log("--------> Replacing command with shim ------>", cmd, binaryPathMap[cmd]);
+      cmd = binaryPathMap[cmd];
+    }    
+
+  return cmd;
+}
+
+// Extend Goosed with a new system configuration
+export const extendGoosed = async (config: ExtensionConfig) => {
+  // allowlist the CMD for stdio type
+  if (config.type === "stdio") {
+    const allowedCMDs = ['goosed', 'npx', 'uvx'];
```

Comment:
nit: I'm sure this was just replicated from the previous version, but i'd remove this check before we forget, don't see why we'd necessarily block an expert building a local system etc

Reply:
yeah if it already asks the user it is probably ok (it is more that it can be triggered from a deep link which opens another risky surface)

Reply:
Good points on both sides! Will leave this check in until we've had a chance to discuss security implications of removing it.